New docs draft (#1040)
near/docs: docs/2.develop/3.development/anatomy/anatomy.md

---
id: anatomy
title: Anatomy of a Contract
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

import MainAs from "./example/main.as.md";
import ModelAs from "./example/models.as.md";

import MainRs from "./example/main.rs.md";
import ModelRs from "./example/models.rs.md";

> When writting smart contracts you will leverage programming concepts such as types, collections, modules, interfaces, and objects among others. Depending on the language you choose to use, the implementation will variate a little. However, the main anatomy of a smart contract follows the same ideas despite how you choose to implement them.

Author: ```suggestion
When writing smart contracts you will leverage programming concepts such as types, collections, modules, interfaces, and objects among others. Depending on the language you choose to use, the implementation will variate a little. However, the main anatomy of a smart contract follows the same ideas despite how you choose to implement them.
```

---

## Example: Anatomy of a Donation Contract

> Lets take a look to the anatomy of a simple contract, which main purpose is to enable donating money to someone. Particulary, the contract defines a `beneficiary` account on initialization, and exposes a method `donation` to forward money while keeping track of it. Please notice that this contract is written for educational purposes only.

Author: ```suggestion
Let's take a look at the anatomy of a simple contract, which main purpose is to enable donating money to someone. Particularly, the contract defines a `beneficiary` account on initialization and exposes a `donation` method to forward money while keeping track of it. Please notice that this contract is written for educational purposes only.
```

:::tip
In RUST we are also relying in the `env::storage` object to store the contract's attributes. However, this gets abstracted away by the SDK.
:::

:::warning
In Assemblyscript you could call multiple times the `init` function. Please make sure to add a check so the function is not called more than once. This can be done by simply adding the following code:

```ts
  const initialized: bool = storage.getPrimitive<bool>('initialized', false)
  assert(!initialized, "Already initialized")
  storage.set<bool>('initialized', true)
```
:::

---

### Classes and NEAR Bindgen

You might have notice in the examples that the classes are decorated with `nearbindgen`:

<Tabs className="language-tabs">
  <TabItem value="as" label="🚀 - Assemblyscript">

  ```ts
    @nearbindgen
    class Donation{
      account_id: string,
      amount: u128
    }
  ```

  </TabItem>
  <TabItem value="rs" label="🦀 - Rust">

  ```rust
    #[near_bindgen]
    #[derive(Default, BorshDeserialize, BorshSerialize)]
    pub struct Donation {
      account_id: AccountId,
      amount: Balance
    }
  ```

  </TabItem>
</Tabs>

> The `nearbindgen` decorator adds methods to the classes, so once instantiated they can be correctly serialized in the contract's state (storage). For example, the class decorated in Assemblyscript originally has only two attributes: `account_id` and `amount`. However, after being decorated it gains methods to be serialized into JSON. This is helpfull since contracts actually obtain and return values encoded in JSON!.

Author: ```suggestion
The `nearbindgen` decorator adds methods to the classes, so once instantiated they can be correctly serialized in the contract's state (storage). For example, the class decorated in Assemblyscript originally has only two attributes: `account_id` and `amount`. However, after being decorated it gains methods to be serialized into JSON. This is helpful since contracts actually obtain and return values encoded in JSON!.
```
